js/driver/virtio/blk.js

```diff
@@ -0,0 +1,141 @@
+// Copyright 2016-present runtime.js project authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+'use strict';
+const VirtioDevice = require('./device');
+const runtime = require('../../core');
+const { Uint64LE } = require('int64-buffer');
+
+const VIRTIO_BLK_T_IN = 0;
+const VIRTIO_BLK_T_OUT = 1;
+
+function initializeBlockDevice(pciDevice) {
+  const ioSpace = pciDevice.getBAR(0).resource;
+  const irq = pciDevice.getIRQ();
+
+  const features = {
+    VIRTIO_BLK_F_SIZE_MAX: 1,
+    VIRTIO_BLK_F_SEG_MAX: 2,
+    VIRTIO_BLK_F_GEOMETRY: 4,
+    VIRTIO_BLK_F_RO: 5,
+    VIRTIO_BLK_F_BLK_SIZE: 6,
+    VIRTIO_BLK_F_FLUSH: 9,
+    VIRTIO_BLK_F_TOPOLOGY: 10,
+    VIRTIO_BLK_F_CONFIG_WCE: 11,
+  };
+
+  const dev = new VirtioDevice('blk', ioSpace);
+  dev.setDriverAck();
+
+  const driverFeatures = {
+    // VIRTIO_BLK_F_SIZE_MAX: true,
+    VIRTIO_BLK_F_SEG_MAX: true,
+    VIRTIO_BLK_F_GEOMETRY: true,
+    VIRTIO_BLK_F_BLK_SIZE: true,
+    VIRTIO_BLK_F_TOPOLOGY: true,
+  };
+
+  const deviceFeatures = dev.readDeviceFeatures(features);
+  debug(JSON.stringify(deviceFeatures));
+
+  if (!dev.writeGuestFeatures(features, driverFeatures, deviceFeatures)) {
+    debug('[virtio] blk driver is unable to start');
+    return;
+  }
+
+  const QUEUE_ID_REQ = 0;
+
+  const reqQueue = dev.queueSetup(QUEUE_ID_REQ);
+  const promiseQueue = [];
+
+  const sectorSize = 512;
+  const sectorCount = dev.blkReadSectorCount();
+  const totalSectorCount = dev.blkReadTotalSectorCount();
+
+  function buildHeader(type, sector) {
+    const u8 = new Uint8Array(16);
+    const view = new DataView(u8.buffer);
+    view.setUint32(0, type, true);
+    view.setUint32(4, 0, true); // priority: low
+    u8.set((new Uint64LE(sector)).toArray(), 8);
+    return u8;
+  }
+
+  const diskDriver = new runtime.disk.DiskInterface('hda', {
```

Comment:
I would do a hierarchy based on the buses (ide, s-ata, usb, i2c...) showing
the real hardware relationships with the machine as namespaces, and later
add us or the user or a detection function some convenience assigments like
hardlinks on filesystems. For example, have `devices['s-ata'][0]` and the
user would do if he want `devices.hda = devices['s-ata'][0]` to access it
easier, and now there are two ways to access to the same device.

El 28/8/2016 9:06, "piranna@gmail.com" <piranna@gmail.com> escribió:

> Maybe we could remove all the hdX thing, each OS has its own way to define
> interfaces. Why not use a Javascript objects hierarchy, like disks[0]? hda
> would be an alias, but not mandatory... Why to use a string when we have
> direct access to the device object?
>
> El 28/8/2016 5:10, "Serge" <notifications@github.com> escribió:
>
>> In js/driver/virtio/blk.js
>> <https://github.com/runtimejs/runtime/pull/133#discussion_r76525427>:
>>
>> > +  const promiseQueue = [];
>> > +
>> > +  const sectorSize = 512;
>> > +  const sectorCount = dev.blkReadSectorCount();
>> > +  const totalSectorCount = dev.blkReadTotalSectorCount();
>> > +
>> > +  function buildHeader(type, sector) {
>> > +    const u8 = new Uint8Array(16);
>> > +    const view = new DataView(u8.buffer);
>> > +    view.setUint32(0, type, true);
>> > +    view.setUint32(4, 0, true); // priority: low
>> > +    u8.set((new Uint64LE(sector)).toArray(), 8);
>> > +    return u8;
>> > +  }
>> > +
>> > +  const diskDriver = new runtime.disk.DiskInterface('hda', {
>>
>> hda name should be assigned by registerDisk I think
>>
>> —
>> You are receiving this because you commented.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/runtimejs/runtime/pull/133/files/4cc7b1fc508923bbc3067a9cb391c465bdac617b#r76525427>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AAgfvszrchC_tZZVlY6GyMKrQyJ2QRFpks5qkPwfgaJpZM4Jur73>
>> .
>>
>